### DISCUSSION TITLE
Relay 4 ghost switching with diy board

### DISCUSSION THREAD

**Discussion by rj10rd** [arendst/Tasmota#18889] [Q&A]
Hello guys, I made my custom PCB with esp32, 4relay, and Rfreceiver. I am using the latest Tasmota 12.5.0 There is an issue sometimes when I turn on/off relay no. 4 (on GPIO 27) It restarts tasmota
Restart reason-  vbat power on reset
I am using tasomocompiler to compile tasmota with following parameters:

#ifdef STA_SSID1 
  #undef STA_SSID1 
#endif 
#define STA_SSID1 "AirFiber" 
 
#ifdef STA_PASS1 
  #undef STA_PASS1 
#endif 
#define STA_PASS1 "airfiber@#" 
 
#ifdef STA_SSID2 
  #undef STA_SSID2 
#endif 
#define STA_SSID2 "Redmi K30" 
 
#ifdef STA_PASS2 
  #undef STA_PASS2 
#endif 
#define STA_PASS2 "akdk200@#" 
 
#ifdef CODE_IMAGE_STR 
  #undef CODE_IMAGE_STR 
#endif 
#define CODE_IMAGE_STR "DK" 
 
#define USER_TEMPLATE    "{\"NAME\":\"ESP32-DevKit\",\"GPIO\":[1,1,1,1,1888,1,1,1,225,224,226,1,1856,480,1152,1,0,1,1,1,0,161,160,227,0,0,0,0,163,162,1,1,1,0,0,1],\"FLAG\":0,\"BASE\":1}" 
 
#ifdef MODULE 
#undef MODULE 
#endif 
#define MODULE                 USER_MODULE 
 
#ifdef USER_RULE1 
#undef USER_RULE1 
#endif 
#define USER_RULE1 "ON Power1#boot DO var1 %value% ENDON ON Power2#boot DO var2 %value% ENDON ON Power3#boot DO var3 %value% ENDON ON power1#state DO var1 %value% ENDON ON power2#state DO var2 %value% ENDON ON power3#state DO var3 %value% ENDON ON RfReceived#Data=0xF9EA68 DO power1 1 ENDON ON RfReceived#Data=0xF9EA62 DO backlog power2 1;power1 0 ENDON ON RfReceived#Data=0xF9EA61 DO power4 toggle ENDON ON RfReceived#Data=0xF9EA64 DO IF ((%var1%==1) OR (%var2%==1) OR (%var3%==1)) backlog power1 0;power2 0;power3 0 ELSEIF ((%var1%==0) AND (%var2%==0) AND (%var3%==0)) power3 1 ENDIF ENDON ON Clock#Timer=1 do backlog power1 0; power2 0; power3 0; power4 0 endon on Tele-SR04#Distance DO backlog var4=((60-%Value%)/2.54)*(23*27)/(61);var5=((var4)/90)*(100) Endon ON event#wp do publish tele/wc/LITERS %var4% ENDON ON event#wp do publish tele/wc/percentage %var5% ENDON" 
 
#ifdef USER_RULE2 
#undef USER_RULE2 
#endif 
#define USER_RULE2 "ON Switch1#state do power1 %value% ENDON ON SWITCH2#state DO POWER2 %VALUE% ENDON ON SWITCH3#state DO POWER3 %VALUE% ENDON ON SWITCH4#state DO POWER4 %VALUE% ENDON ON SR04#DISTANCE>60 DO backlog power4 0;buzzer 1,5 ENDON ON event#wp do publish tele/wc/LITERS %var4% ENDON ON event#wp do publish tele/wc/percentage %var5% ENDON" 
 
#ifdef USE_EXPRESSION 
#undef USE_EXPRESSION 
#endif 
#define USE_EXPRESSION 
 
#ifdef SUPPORT_IF_STATEMENT 
#undef SUPPORT_IF_STATEMENT 
#endif 
#define SUPPORT_IF_STATEMENT 
 
 
#ifdef USER_BACKLOG 
#undef USER_BACKLOG 
#endif 
#define USER_BACKLOG "Rule1 1;Rule2 1;Timezone +5:30;switchmode1 2;switchmode2 2;switchmode3 2;switchmode4 2;RfProtocol 1;setoption55 1;Rule2 5;event wp" 
 
#ifdef MQTT_USER 
  #undef MQTT_USER 
#endif 
#define MQTT_USER              "xxxx" 
 
#ifdef MQTT_HOST 
  #undef MQTT_HOST 
#endif 
#define MQTT_HOST           "xxxxxxxxxxx" 
 
#ifdef MQTT_PORT 
  #undef MQTT_PORT 
#endif 
#define MQTT_PORT              8883 
 
#ifdef MQTT_PASS 
  #undef MQTT_PASS 
#endif 
#define MQTT_PASS              "xxxxxx" 
 
 
#ifdef MQTT_TOPIC 
  #undef MQTT_TOPIC 
#endif 
#define MQTT_TOPIC             "wc" 
 
 
#ifdef FRIENDLY_NAME 
  #undef FRIENDLY_NAME 
#endif 
#define FRIENDLY_NAME          "Water Cooler" 
 
#undef USE_BERRY 
#undef USE_AUTOCONF

**Comment by sfromis** [comment]
I'd guess that what you experience is more related to hardware, and "something" triggering boots. Maybe something bad with the wiring about relay4. To try to debug more, I'd recommend you to run with serial connected and follow the serial console around the point of restart, probably after having activated `Seriallog 3` for details. But hardware issues are generally not easy to debug from looking at the software level.

Anyway, Instead of deactivating Berry, I'd really recommend you to use Berry to avoid all of that messing with complexity in rules. You get much better features for doing programming.

**Reply by rj10rd** [reply]
Thanks for the reply. I am a noob at coding. Can you please suggest a place to learn berry from Strach?
BTW, I am using relays on the following pins:
GPIO12
GPIO13
GPIO14
GPIO27
And I am also using induction loads; can this cause problems? I can also share my PCB schematic if it helps in finding the cause.

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/Berry/

What gpios you use does probably not make a difference, more like what comes next, like what drivers you have for the relays.
My electronics debugging knowledge is rudimentary, but maybe someone else can spot something....

**Reply by rj10rd** [reply]
[pcb.pdf](https://github.com/arendst/Tasmota/files/11773263/pcb.pdf)
My PCB Schematic

**Reply by sfromis** [reply]
Did you try my suggestion of checking serial output at the time of the restarts?

**Reply by barbudor** [reply]
Your schematic looks electrically correct although barely readable. Which makes me wonder how bad your PCB routing could be.

Having the ESP resetted when enabling a relay is typical of bad routing, lack of proper power separation. Probably current peaks going through too thin tracks and too close to the ESP.

**Reply by rj10rd** [reply]
Thanks for the reply, sir. ESP doesn't get reset; it just restarts sometimes. I am also attaching a Gerber file for any suggestions or improvements I can make to the PCB. For the serial log debugging, I am unable to do it now but will surely upload logs soon.
[pcbo.zip](https://github.com/arendst/Tasmota/files/11779377/pcbo.zip)

**Reply by sfromis** [reply]
"Restarts sometimes" is another way of saying "getting reset".

**Reply by rj10rd** [reply]
Thanks for the reply. Is a pullup resistor required for buttons? Because after removing buttons from the esp32, my problem is solved. BTW, there is nothing in the serial logs.

**Reply by sfromis** [reply]
With normal setup (buttons connecting a pin to ground when pushed), the internal pullup resistors automatically enabled will do the job, and no "surprises". But if a button registers as "pushed" for 40 seconds (which can happen with a config not matching the hardware), that enables a default recovery feature to cause a full "factory reset". And you should get a resetting message in the serial log, if it is active when it happens.
